### ISSUE TITLE
NoSuchMethodError: The getter 'reactiveSingleton' was called on null

### ISSUE THREAD

**Issue by SGarno** [GIfatahTH/states_rebuilder#29]
I have just converted to the latest release (1.8.0+1) and am now getting a lot of the above errors.

Nothing has changed in my code other than the use of getAsReactive instead of getAsModel.

Here is a trace of the error

```
I/flutter ( 6949): Instance of 'StreamStatesRebuilder<AppConfig>'
E/flutter ( 6949): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: NoSuchMethodError: The getter 'reactiveSingleton' was called on null.
E/flutter ( 6949): Receiver: null
E/flutter ( 6949): Tried calling: reactiveSingleton
E/flutter ( 6949): #0      Object.noSuchMethod  (dart:core-patch/object_patch.dart:53:5)
E/flutter ( 6949): #1      ReactiveModel._rebuildStates 
package:states_rebuilder/src/reactive_model.dart:179
E/flutter ( 6949): #2      ReactiveModel.setState 
package:states_rebuilder/src/reactive_model.dart:164
E/flutter ( 6949): #3      HomeScreen._buildItem.<anonymous closure>.<anonymous closure> 
package:rapidlog/screens/HomeScreen.dart:87
E/flutter ( 6949): #4      _rootRunUnary  (dart:async/zone.dart:1134:38)
```

Here is the code I am using in one of the functions:

```
              final config = Injector.getAsReactive<AppConfig>();
              config.setState( (state) => state.add(newLogger) );
```

It is behaving like a timing issue.    Sometimes when I run the debugger and put a breakpoint on the first line, and step through it, config returns the value.   Other times, config is not defined, and explicitly attempting to get its value in the debug console, I get the following error:

```
'': error: org-dartlang-debug:synthetic_debug_expression:1:1: Error: The getter 'config' isn't defined for the class 'LibraryListScreen'.
 - 'LibraryListScreen' is from 'package:rapidlog/screens/LibraryListScreen.dart' ('lib/screens/LibraryListScreen.dart').
Try correcting the name to the name of an existing getter, or defining a getter or field named 'config'.
```

I read through the change log and saw the introduction of getAsReactive, and made the appropriate changes, but is there something else that I need to change in my code with this new release?   (I was previously using 1.6.1)

**Comment by GIfatahTH** [comment]
Thank you for reporting this issue.

Form what i expect: 
You are injecting a stream using `Inject.stream` and in :
```dart
final config = Injector.getAsReactive<AppConfig>();
config.setState( (state) => state.add(newLogger) );
```
you are getting the injected stream as reactive and call setState.

This is not allowed. By injecting a stream using `Inject.stream` you are automatically listening to it, whenever the stream emits a value all listeners will be notified with the emitted value.

so config.setState is not allowed.

try injecting the AppConfig with `Inject` without named stream constructor.

**Comment by GIfatahTH** [comment]
Here you find an example on how to use streams.
https://github.com/GIfatahTH/states-rebuilder-examples/tree/master/003-login_form_with_validation_using_streams

**Comment by SGarno** [comment]
Thank you for your reply. 

I am not using streams, however, I am using futures; so I am assuming they behave the same?

Here is my code:
```
    return Injector(
      inject: [
        Inject<AppConfig>.future( () => AppConfig().load() ),
        Inject<Library>.future( () => Library().load() ),
      ],
      // appLifeCycle: ( state ) {
      //   if( state != AppLifecycleState.resumed ) {
      //     final config = Injector.getAsReactive<AppConfig>().state;
      //     config.save(config);
      //   }
      // },
      builder: (context) {

        final configSnap = Injector.getAsReactive<AppConfig>(context: context).snapshot;
        final librarySnap = Injector.getAsReactive<Library>(context: context).snapshot;

        if( configSnap.hasData == false || librarySnap.hasData == false ) {
          return MaterialApp( 
            home: SplashScreen()
          );
        }

        return MaterialApp(
          title: configSnap.data.appName,
          theme: ThemeData(
            primarySwatch: Colors.blue,
          ),
          home: Material(
            child: HomeScreen()
          )
        );
      }
    );

```

BTW:  The link you provided is not accessible (404).

**Comment by SGarno** [comment]
Some additional feedback....

I looked at how you deal with ConnectionState in your counter app with future example, and updated the code accordingly, just to see if that is what was going on.

Here is my code:

```
            _showEditLogger(context,logger).then( (newLogger) {
              final config = Injector.getAsReactive<AppConfig>();
              print('ConnectionState ${config.connectionState}');
              if( config.connectionState != ConnectionState.waiting ) {
                config.setState( (state) => state.add(newLogger) );
                Navigator.pop(context);
              }
            });
```

I still get the error, yet the connection state is done.

```
I/flutter ( 7667): ConnectionState ConnectionState.done
E/flutter ( 7667): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: NoSuchMethodError: The getter 'reactiveSingleton' was called on null.
E/flutter ( 7667): Receiver: null
E/flutter ( 7667): Tried calling: reactiveSingleton
```

I am going to refactor my code to remove a future injector.   I just thought it was a cool way to take care of loading a config file, but if I now have to monitor connectionState everywhere, that complicates the code.

Anyway, still not sure why the error is occurring if the connectionState is 'done'.

**Comment by GIfatahTH** [comment]
Is this kind of code resemble yours:

Try it and you see that it works. But there is a better way to get the same result.

The link works for me! It is my new repos. this is the [link](https://github.com/GIfatahTH/states-rebuilder-examples)
```dart
import 'package:flutter/material.dart';
import 'package:states_rebuilder/states_rebuilder.dart';

class AppConfig {
  String appName = 'appName';
  Future<AppConfig> load() async {
    await Future.delayed(Duration(seconds: 1));
    print('AppConfig loaded');
    return AppConfig();
  }
}

class Library {
  Future<Library> load() async {
    await Future.delayed(Duration(seconds: 2));
    print('Library loaded');
    return Library();
  }
}

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Injector(
          inject: [
            Inject<AppConfig>.future(() => AppConfig().load()),
            Inject<Library>.future(() => Library().load()),
          ],
          builder: (context) {
            final configSnap =
                Injector.getAsReactive<AppConfig>(context: context).snapshot;
            final librarySnap =
                Injector.getAsReactive<Library>(context: context).snapshot;

            if (configSnap.hasData == false || librarySnap.hasData == false) {
              return MaterialApp(
                home: Center(child: Text('SplashScreen')),
              );
            }

            return MaterialApp(
                title: configSnap.data.appName,
                theme: ThemeData(
                  primarySwatch: Colors.blue,
                ),
                home: Material(child: Center(child: Text('HomeScreen'))));
          },
        ),
      ),
    );
  }
}
```

**Comment by SGarno** [comment]
Yes, what you have there is indeed similar, but I wonder if it has something to do with a deep object that then gets used later in the application as opposed to just a simple example as you have constructed here.   Not sure, because it also worked on a previous version 1.6.1 without issue.

I have since refactored my code to not use future and now everything is fine again.

```
  _load(AppConfig config, Library library) async {
    await config.load();
    await library.load();

    setState(() {
      isLoaded = true;
    });

    return( [config, library] );
  }

    return Injector(
      inject: [
        Inject<AppConfig>( () => AppConfig() ),
        Inject<Library>( () => Library() ),
      ],
      // appLifeCycle: ( state ) {
      //   if( state != AppLifecycleState.resumed ) {
      //     final config = Injector.getAsReactive<AppConfig>().state;
      //     config.save(config);
      //   }
      // },
      builder: (context) {

        var config = Injector.get<AppConfig>();
        var library = Injector.get<Library>();

        if( !isLoaded ) {
          _load(config, library);
          print('Application loaded');
        }

        return MaterialApp(
          title: config.appName,
          theme: ThemeData(
            primarySwatch: Colors.blue,
          ),
          home: Material(
            child: isLoaded ? HomeScreen() : SplashScreen()
          )
        );
      });
  }
```

**Comment by GIfatahTH** [comment]
Look at this:
Always keep your model simple without any reference to anything that has a relation with the UI.

This a prototype of the models of your case:

```dart
class AppConfig {
  String appName = 'appName';
  load() async {
    await Future.delayed(Duration(seconds: 1));
    print('AppConfig loaded');
  }
}

class Library {
  load() async {
    await Future.delayed(Duration(seconds: 2));
    print('Library loaded');
  }
}
```
Then you have to inject your models in the widget tree where you need them.
In the code, in the initState of the `Injector` widget, I get the reactive models and call load method.

```dart

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Injector(
          //Inject the model not the future
          inject: [
            Inject<AppConfig>(() => AppConfig()),
            Inject<Library>(() => Library()),
          ],
          initState: () {
          // call load methods
            Injector.getAsReactive<AppConfig>()
                .setState((state) => state.load());
            Injector.getAsReactive<Library>().setState((state) => state.load());
          },
          builder: (context) {
           // get the models not the snpashots
          // The models are obtained with the context parameter. 
            final configModel =
                Injector.getAsReactive<AppConfig>(context: context);
            final libraryMdoel =
                Injector.getAsReactive<Library>(context: context);

            if (configSnap.hasData == false || librarySnap.hasData == false) {
              return MaterialApp(
                home: Center(child: Text('SplashScreen')),
              );
            }

            return MaterialApp(
                title: configSnap.state.appName,
                theme: ThemeData(
                  primarySwatch: Colors.blue,
                ),
                home: Material(child: Center(child: Text('HomeScreen'))));
          },
        ),
      ),
    );
  }
}
```

**Comment by GIfatahTH** [comment]
another approach is :

```dart
import 'package:flutter/material.dart';
import 'package:states_rebuilder/states_rebuilder.dart';

class AppConfig {
  String appName = 'appName';
  load() async {
    await Future.delayed(Duration(seconds: 1));
    print('AppConfig loaded');
  }
}

class Library {
  load() async {
    await Future.delayed(Duration(seconds: 2));
    print('Library loaded');
  }
}

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Injector(
          inject: [
            Inject<AppConfig>(() => AppConfig()),
            Inject<Library>(() => Library()),
          ],
          builder: (context) {
             //The models are obtained without the context, subscription is done using StateBuilder widget
            final configModel = Injector.getAsReactive<AppConfig>();
            final libraryModel = Injector.getAsReactive<Library>();

            return StateBuilder(
              models: [configModel, libraryModel],
              initState: (_, __) {
                configModel.setState((state) => state.load());
                libraryModel.setState((state) => state.load());
              },
              builder: (_, __) {
                if (configModel.hasData == false ||
                    libraryModel.hasData == false) {
                  return MaterialApp(
                    home: Center(child: Text('SplashScreen')),
                  );
                }

                return MaterialApp(
                  title: configModel.state.appName,
                  theme: ThemeData(
                    primarySwatch: Colors.blue,
                  ),
                  home: Material(
                    child: Center(
                      child: Text('HomeScreen'),
                    ),
                  ),
                );
              },
            );
          },
        ),
      ),
    );
  }
}
```

**Comment by GIfatahTH** [comment]
What about the [link](https://github.com/GIfatahTH/states-rebuilder-examples). Did it work for you?

**Comment by SGarno** [comment]
Yes, I am keeping no UI within the models.   Only in the widgets.

I like the second one, very clean!   

I think the part I missed was the fact that StateBuilder had initState which could still be used in a stateless widget.   I was extrapolating that to being required in a stateful widget.   I will refactor and let you know how that works.    

As for the link, no, it still does not work.   I get a 404.     Try hitting the link using an incognito tab or some other browser.   I cannot see it.

**Comment by GIfatahTH** [comment]
The [link ](https://github.com/GIfatahTH/states-rebuilder-examples) works now.

**Comment by GIfatahTH** [comment]
Yes, in addition to its state management responsibility, `StateBuilder` offers a facade that facilitates the management of the widgets lifecycle.
```dart
StateBuilder<T>(
  onSetState: (BuildContext context, ReactiveModel<T> model){
  /*
  Side effects to be executed after sending notification and before rebuilding the observers. Side effects are navigating, opening the drawer, showing snackBar,...  
  
  It is similar to 'onSetState' parameter of the 'setState' method. The difference is that the `onSetState` of 'setState' method is called once after executing the 'setState'. But this 'onSetState' is executed each time a notification is send from one of the observable model this 'StateBuilder' is subscribing.
  
  You can use another nested setState here.
  */
  },
  onRebuildState: (BuildContext context, ReactiveModel<T> model){
  // The same as in onSetState but called after the end rebuild process.
  },
  initState: (BuildContext context, ReactiveModel<T> model){
  // Function to execute in initState of the state.
  },
  dispose: (BuildContext context, ReactiveModel<T> model){
  // Function to execute in dispose of the state.
  },
  didChangeDependencies: (BuildContext context, ReactiveModel<T> model){
  // Function to be executed  when a dependency of state changes.
  },
  didUpdateWidget: (BuildContext context, ReactiveModel<T> model, StateBuilder oldWidget){
  // Called whenever the widget configuration changes.
  },
  afterInitialBuild: (BuildContext context, ReactiveModel<T> model){
  // Called after the widget is first inserted in the widget tree.
  },
  afterRebuild: (BuildContext context, ReactiveModel<T> model){
  /*
  Called after each rebuild of the widget.

  The difference between onRebuildState and afterRebuild is that the latter is called each time the widget rebuilds, regardless of the origin of the rebuild. 
  Whereas onRebuildState is called only after rebuilds after notifications from the models to which the widget is subscribed.
  */
  },
  // If true all model will be disposed when the widget is removed from the widget tree
  disposeModels: true,

  // A list of observable objects to which this widget will subscribe.
  models: [model1, model2]

  // Tag to be used to filer notification from observable classes.
  // It can be any type of data, but when it is a List, 
  // this widget will be saved with many tags that are the items in the list.
  tag: dynamic

  builder: (BuildContext context, ReactiveModel<T> model){
    /// [BuildContext] can be used as the default tag of this widget.

    /// The model is the first instance (model1) in the list of the [models] parameter.
    /// If the parameter [models] is not provided then the model will be a new reactive instance.
  },
  builderWithChild: (BuildContext context, ReactiveModel<T> model, Widget child){
    ///Same as [builder], but can take a child widget containing the part of the widget tree that we do not want to rebuild.
    /// If both [builder] and [builderWithChild] are defined, it will throw.

  },

  //The child widget that is used in [builderWithChild].
  child: MyWidget(),

)
```

**Comment by SGarno** [comment]
I really love what you have done here!   I think this should have been done to begin with in Flutter.   

There are some widgets, though, that need to be completely re-usable with no understanding of the application environment.   In those cases, it still seems that stateful widgets are required (as posted in issue #27) or perhaps do some sort of channel-type approach.

Can you put the Injector in a widget that the app also has an Injector?   Maybe I can see if that is a solution (assuming StatesRebuilder overhead is minimal).

**Comment by GIfatahTH** [comment]
>I really love what you have done here! I think this should have been done to begin with in Flutter.

Very happy with that! Thank you.

> In those cases, it still seems that stateful widgets are required
Yes, there are still some cases where the use of a stateful widget is required. But for most practical cases `StateBuilder` and `StateWithMixinBuilder` do a good job.

>Can you put the Injector in a widget that the app also has an Injector? Maybe I can see if that is a solution (assuming StatesRebuilder overhead is minimal).

Can you explain more? If I understand well, I say yes you can use nested Injectors.

**Comment by SGarno** [comment]
I will put my response in #27 regarding stateful widgets....it seems more appropriate there.   This was about the error....which I am not sure is resolved, I just refactored it away, but it may be worth investigating if it shows up again?

**Comment by GIfatahTH** [comment]
The error is expected because you can not use the setState method on a model injected with `Inject.stream` of 'Inject.future`. In the case of models injected using  `Inject.stream` of 'Inject.future` the setState method is called automatically when the stream emits a value or the future completes. In older versions, This was allowed but from version 1.8.0 this is not allowed. I will add an assertion message in the next version.

**Comment by SGarno** [comment]
Ah!  That explains it.   Yes, an assert would be good to have so that it is clear it is not just an error with coding, but something that needs to be addressed in the code.

**Comment by SGarno** [comment]
Well, after refactoring my code a bit...it is now much cleaner (thank you!)

Here is what it looks like now as a result of having converted it not use future (ala initState) which also allowed me to make it a stateless widget!   Awesome.    

```
class MyApp extends StatelessWidget {

  @override
  Widget build(BuildContext context) {

    return Injector(
      inject: [
        Inject<Library>( () => Library() ),
        Inject<AppConfig>( () => AppConfig() ),
      ],
      initState: () {
        _loadLibrary();
        _loadConfig();
      },
      builder: (context) {
        final lib = Injector.getAsReactive<Library>(context: context);
        final config = Injector.getAsReactive<AppConfig>(context: context);
        final isLoaded = lib.hasData && config.hasData;

        print( isLoaded ? "Data is loaded." : "No data yet.." );

        return MaterialApp(
          title: config.state?.appName ?? "MyApp",
          theme: ThemeData(
            primarySwatch: Colors.blue,
          ),
          home: Material(
            child: isLoaded ? HomeScreen() : SplashScreen()
          ),
        );
      }
    );
  }
}
```

**Comment by SGarno** [comment]
A bit of a snag.....This is so nice and clean, but I realized that this requires the _loadLibrary() function to load into the instance previously Injected.    When coming from a json repository, I prefer to make the model members final and the fromJson function a static constructor.

This creates a problem with the above code because my _loadLibrary function creates a new Library object and somehow needs to re-inject that into StatesRebuilder.

Without reverting back to a stateful widget or using futures, how can this be accomplished?  Is there a way to replace an existing model instance in StatesRebuilder?

**Comment by GIfatahTH** [comment]
Can you show me how your Library class looks like?

**Comment by SGarno** [comment]
Yes, it is just a plain ol Dart class.

```
class Library {
  String version;
  List<LibraryItem> items;

  Library({this.items, this.version = LIBRARY_VERSION});

  fromJson(Map<String, dynamic> json) {
    this.version = json['version'];
    
    if (json['items'] != null) {
      this.items = List<LibraryItem>();
      json['items'].forEach((v) {
        this.items.add(LibraryItem.fromJson(v));
      });
    }    
  
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();

    data['version'] = LIBRARY_VERSION;

    if (this.items != null) {
      data['items'] = this.items.map((v) => v.toJson()).toList();
    }

    return data;
  }

}
```

I would prefer that all the members of the class were final and the `fromJson` function was static, but that creates a problem with the injector pattern because I wont have the instance yet until the load is performed; and there are a few things that need to happen to determine how that library is loaded. 

Here is my injector:

```
    return Injector(
      inject: [
        Inject<Library>( () => Library() ),
      ],
      initState: () {
        _loadLibrary();
      },
```

and the _loadLibrary function:

```
  _loadLibrary() async {

    var hasPermissions = await _hasFilePermission();
    var lib = Injector.getAsReactive<Library>();

    Map<String,dynamic> json;

    // Try to load from the file
    if( hasPermissions ) {
      json = await JsonFileRepository(LIBRARY_FILENAME).load(); 
      json ?? print( 'Loaded library file $LIBRARY_FILENAME');
    }

    // If that fails, then load from the embedded asset
    if( json == null ) {
      json = await JsonAssetRepository(LIBRARY_ASSET).load(); 
      json ?? print( 'Loaded asset file $LIBRARY_ASSET');
    }

    if( json != null ) {
      lib.setState( (state) => lib.state.fromJson(json));
    }
  }
```

I know I can refactor this so that the load occurs prior to the Injector (that is how I had it previously), but I really like the standard injector pattern where I only load in the initState.    Using a future injector requires a lot more boilerplate to handle in all the widgets; so if I can keep it simple, that is my preference.    For now, I am sticking with a plain non-final class, but if there is a better way, that would be great.

**Comment by GIfatahTH** [comment]
I was expecting that. The Library is a data model class with fields and fromJson and toMap (toJson) methods. DATA model class should never injected to the UI because they are expected to be instantiated many times during app life. Instead consider injection a service (use case) class which has the responsibility of creating the Library class (through repository) and do CRUD operation.

Try to move _loadLibrary out of the UI to another plain dart class. This new class is the service (use case) class. Then inject this class and call the _loadLibrary using serState. 
Btw. I am preparing a series of articles on implementing the clean architecture using States_rebuilder

**Comment by SGarno** [comment]
That would be great!   Would love to see how you recommend it to be constructed.

Yes, I do agree that the LoadLibrary should not be in the UI, but considering it is in initState, wouldn't that only be called once?

This widget is the first widget of the app, the only other place to put this would then have to be in the main.dart.   All the other widgets do not use this approach, in fact, I am using EventBus to separate the business logic outside of the widgets in a much cleaner fashion.   Somewhat along the lines of BLoC architecture, but a bit more simplified.   I am still using Events and Models, but trying to eliminate the BLoC delegates and some of the boilerplate code that makes BLoC a bit more tedious.

Let me know when you have that architecture guide published, I would be interested in seeing the appropriate way to initialize a JSON based configuration/library into an app.    If you had an EventBus approach as well, I would be all over that!

Thanks again!

**Comment by GIfatahTH** [comment]
issue fixed in 1.11.0 release.
